src/paymentdisclosure.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2017 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "paymentdisclosure.h"
+#include "util.h"
+
+std::string PaymentDisclosureInfo::ToString() const {
+    return strprintf("PaymentDisclosureInfo(version=%d, esk=%s, joinSplitPrivKey=%s, address=%s)",
+        version, esk.ToString(), joinSplitPrivKey.ToString(), CZCPaymentAddress(zaddr).ToString());
+}
+
+std::string PaymentDisclosure::ToString() const {
+    std::string s = HexStr(payloadSig.begin(), payloadSig.end());
+    return strprintf("PaymentDisclosure(payload=%s, payloadSig=%s)", payload.ToString(), s);
+}
+
+std::string PaymentDisclosurePayload::ToString() const {
+    return strprintf("PaymentDisclosureInfo(version=%d, esk=%s, txid=%s, js=%d, n=%d, address=%s, message=%s)",
+        version, esk.ToString(), txid.ToString(), js, n, CZCPaymentAddress(zaddr).ToString(), message);
+}
+
+PaymentDisclosure::PaymentDisclosure(const uint256 &joinSplitPubKey, const PaymentDisclosureKey &key, const PaymentDisclosureInfo &info, const std::string &message)
+{
+    // Populate payload member variable
+    payload.version = info.version; // experimental = 0, production = 1 etc.
+    payload.esk = info.esk;
+    payload.txid = key.hash;
+    payload.js = key.js;
+    payload.n = key.n;
+    payload.zaddr = info.zaddr;
+    payload.message = message;
+
+    // Serialize and hash the payload to generate a signature
+    uint256 dataToBeSigned = SerializeHash(payload, SER_GETHASH, 0);
```

Comment:
> The payload needs to have an encoding that is disjoint from a transaction encoding.

Could you elaborate on your reasoning? Is there an attack or potential parser issue that this is defending against?

Reply:
To elaborate on my own question: I agree that the proposal about *would* ensure that a payment disclosure signature could not collide with either the basic transaction encoding (used to obtain its txid) or the sighash encoding (used for inputs or `joinSplitSig`). But I don't see why this is necessary - where could a payment disclosure payload be mistaken for a transaction, or vice versa?

Reply:
It's best practice to ensure that whenever different protocol messages can be signed with the same keys, they have disjoint encodings such that an encoding from one context will be rejected in the other. (As it happens that is currently the case for this PR, but only by accident, because 0 is not currently a valid transaction version. It could be a valid transaction version in future, and there may in fact be advantages to having the transaction version decrease in order to ensure a bilateral hard fork for NU0.)

Suppose that a payment disclosure is accepted as a transaction. Then this could potentially break the protection against malleability of JoinSplits by allowing you to replay a JoinSplit proof. (The protocol has some defence-in-depth in that area: you'd have to replay the proof before the original transaction got into the blockchain, or orphan a block, in order to avoid repeated nullifiers from causing the replay to be treated as a double-spend. But it is possible to do that.)

Suppose conversely that a transaction is accepted as a payment disclosure. Then someone could potentially forge a disclosure from my transaction without knowing esk. Actually this attack probably wouldn't work because the format of a transaction is too constrained for the payment disclosure to be valid, but it would be annoyingly hard to rule out.

Reply:
In that case, I'd prefer that we instead define the proposal as a 4-byte magic string or flag, where the most significant bit of the last byte must be set (making it a negative number when parsed as an `int32_t`). That more accurately conveys the intent (as a kind of domain separation).

Reply:
Good discussion.  I'm going with int32 value of -9411486 which is 0xFF706462 as the magic bytes (in ascii, it is ÿpdb, where pdb is payment diclosure blob)

Reply:
Remember that whatever value we choose has to be excluded as a future transaction version number. That's why I suggested INT32_MIN, as being maximally "out of the way" of any such future tx version numbers. But the protocol spec can easily reserve the number you suggested, or possibly all values of the form 0xFFxxxxxx (i.e. {-16777216..-1})